File: previous_project_data/0_ported/celestial_body.rs
use bevy::prelude::*;
use serde::{Deserialize, Serialize};
use uuid::Uuid;
use crate::components::orbital_mechanics::OrbitalState;

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub enum CelestialBodyType {
    Star,
    Planet,
    Moon,
    Asteroid,
    Comet,
    DwarfPlanet,
}

#[derive(Component, Debug, Clone, Deserialize, Serialize)]
pub struct CelestialBody {
    pub name: String,
    pub id: Uuid,
    pub body_type: CelestialBodyType,
    pub orbital_state: Option<OrbitalState>,
    pub mass: f64,  // Mass in kg
    pub diameter: f64,  // Diameter in km
}

impl CelestialBody {
    pub fn new(name: String, body_type: CelestialBodyType, mass: f64, diameter: f64) -> Self {
        Self {
            name,
            id: Uuid::new_v4(),
            body_type,
            orbital_state: None,
            mass,
            diameter,
        }
    }
    
    pub fn with_orbital_state(mut self, orbital_state: OrbitalState) -> Self {
        self.orbital_state = Some(orbital_state);
        self
    }
    
    pub fn update_orbital_position(&mut self, days_elapsed: f64) {
        if let Some(ref mut orbital_state) = self.orbital_state {
            orbital_state.update_position(days_elapsed);
        }
    }
    
    pub fn has_significant_position_change(&self, previous_angle: f64, threshold_degrees: f64) -> bool {
        if let Some(ref orbital_state) = self.orbital_state {
            orbital_state.is_significant_change(previous_angle, threshold_degrees)
        } else {
            false
        }
    }
}

// Marker component for celestial bodies that should be rendered
#[derive(Component)]
pub struct CelestialBodyRenderable;

// Component for tracking previous positions for change detection
#[derive(Component)]
pub struct PreviousPosition {
    pub angle: f64,
}

File: old/Common/CelestialBody.cs
#nullable enable
using System;
using Godot;

namespace HarshRealm.Data;

public enum CelestialBodyType
{
    Star,
    Planet,
    Moon,
    Asteroid,
    Comet,
    DwarfPlanet,
}

[GlobalClass]
public partial class CelestialBody : Node
{
    [Export] public string BodyName { get; set; } = string.Empty;
    [Export] public string BodyId { get; set; } = string.Empty;
    [Export] public CelestialBodyType BodyType { get; set; }
    [Export] public double Mass { get; set; } // Mass in kg
    [Export] public double Diameter { get; set; } // Diameter in km
    
    public OrbitalState? OrbitalState { get; set; }

    public CelestialBody()
    {
        BodyId = Guid.NewGuid().ToString();
    }

    public static CelestialBody Create(string name, CelestialBodyType bodyType, double mass, double diameter)
    {
        var body = new CelestialBody
        {
            BodyName = name,
            BodyId = Guid.NewGuid().ToString(),
            BodyType = bodyType,
            Mass = mass,
            Diameter = diameter
        };
        return body;
    }
    
    public CelestialBody WithOrbitalState(OrbitalState orbitalState)
    {
        OrbitalState = orbitalState;
        return this;
    }
    
    public void UpdateOrbitalPosition(double daysElapsed)
    {
        OrbitalState?.UpdatePosition(daysElapsed);
    }
    
    public bool HasSignificantPositionChange(double previousAngle, double thresholdDegrees)
    {
        return OrbitalState?.IsSignificantChange(previousAngle, thresholdDegrees) ?? false;
    }

    public override string ToString() => $"{BodyName} ({BodyType})";
}

/// <summary>
/// Marker component for celestial bodies that should be rendered
/// </summary>
[GlobalClass]
public partial class CelestialBodyRenderable : Node
{
    // Marker component - no additional properties needed
}

/// <summary>
/// Component for tracking previous positions for change detection
/// </summary>
[GlobalClass]
public partial class PreviousPosition : Node
{
    [Export] public double Angle { get; set; }
    
    public PreviousPosition() { }
    
    public PreviousPosition(double angle)
    {
        Angle = angle;
    }
}
